README.md
# Embeddings NPM Package

The `embeddings` package is a utility for generating question-answer pairs and embeddings from HTML pages or text input. It utilizes the OpenAI API to generate question-answer pairs and embeddings. This package is useful for generating training data for chatbots or question-answering models.

## Constructor Options

The `Embeddings` class can be instantiated with the following options:

- `apiKey` (required): Your OpenAI API key.
- `embeddingModel` (optional, default: "text-embedding-ada-002"): The name of the OpenAI model to use for generating embeddings.
- `completionModel` (optional, default: "text-davinci-003"): The name of the OpenAI model to use for generating question-answer pairs.
- `completionModelOptions` (optional, default: `{ max_tokens: 2000, n: 1, stop: null, temperature: 0.7 }`): The options to pass to the completion model when generating question-answer pairs.
- `screenshotApiKey` (optional): Your Pagepixels Screenshot API key (https://pagepixels.com), used for scraping HTML from webpages.
- `screenshotOptions` (optional, default: `{}`): The options to pass to the Pagepixels Screenshot API when scraping HTML.
- `chunkMaxTokens` (optional, default: 800): The maximum number of tokens to send to the OpenAI API at once.
- `promptRefinement` (optional, default: ""): Any prompt refinement you would like to add to the completion prompt.
- `verbose` (optional, default: false): Whether or not to output additional logging information during processing.

## Usage

The `Embeddings` class provides several methods for generating embeddings and question-answer pairs. These methods can be used standalone or in combination to generate embeddings and question-answer pairs from HTML pages or text input.

### `generateQaEmbeddingsFromText` Method

The `generateQaEmbeddingsFromText` method takes a string of text and generates embeddings and question-answer pairs from it. The method returns an array of objects, each containing the original question-answer pair along with the corresponding embedding.

```javascript
const Embeddings = require('embeddings');

const options = {
  apiKey: 'your_api_key',
  verbose: true
}

const embeddingsClient = new Embeddings(options);

const text = `Welcome to our documentation. This guide will walk you through the basics of using our platform.`;

const embeddingsResult = await embeddingsClient.generateQaEmbeddingsFromText(text);

console.log(embeddingsResult);
```

### `generateQaEmbeddingsFromUrls` Method

The `generateQaEmbeddingsFromUrls` method takes an array of URLs and generates embeddings and question-answer pairs from the text content of the pages at those URLs. The method takes screenshots of the web pages using the Pagepixels API and extracts the text content from the resulting HTML. The method returns an array of objects, each containing the original question-answer pair along with the corresponding embedding and the URL of the page from which it was generated.

```javascript
const Embeddings = require('embeddings');

const options = {
  apiKey: 'your_api_key',
  screenshotApiKey: 'your_screenshot_api_key',
  verbose: true
}

const embeddingsClient = new Embeddings(options);

const urls = ['https://www.example.com', 'https://www.example.com/about'];

const embeddingsResult = await embeddingsClient.generateQaEmbeddingsFromUrls(urls);

console.log(embeddingsResult);
```

### `generateQaEmbeddingsFromQaPairs` Method

The `generateQaEmbeddingsFromQaPairs` method takes an array of question-answer pairs and generates embeddings for the questions. The method returns an array of objects, each containing the original question-answer pair along with the corresponding embedding.

```javascript
const Embeddings = require('embeddings');

const options = {
  apiKey: 'your_api_key',
  verbose: true
}

const embeddingsClient = new Embeddings(options);

const qaPairs = [
  {
    question: "What is the purpose of this documentation?",
    answer: "To guide users through the basics of using the platform."
  }
];

const embeddingsResult = await embeddingsClient.generateQaEmbeddingsFromQaPairs(qaPairs);

console.log(embeddingsResult);
```

### `generateQaPairsFromText` Method

The `generateQaPairsFromText` method takes a string of text and generates question-answer pairs from it. The method returns an array of objects, each containing the question and the answer.

```javascript
const Embeddings = require('embeddings');

const options = {
  apiKey: 'your_api_key',
  verbose: true
}

const embeddingsClient = new Embeddings(options);

const text = `Welcome to our documentation. This guide will walk you through the basics of using our platform.`;

const qaPairs = await embeddingsClient.generateQaPairsFromText(text);

console.log(qaPairs);
```

### `generateEmbeddingForText` Method

The `generateEmbeddingForText` method takes a string of text and generates an embedding for it. The method returns the generated embedding.

```javascript
const Embeddings = require('embeddings');

const options = {
  apiKey: 'your_api_key',
  verbose: true
}

const embeddingsClient = new Embeddings(options);

const text = "What is the purpose of this documentation?";

const embedding = await embeddingsClient.generateEmbeddingForText(text);

console.log(embedding);
```

## Conclusion

The `embeddings` package provides a convenient way to generate question-answer pairs and embeddings from HTML pages or text input using the OpenAI API. By using the methods provided by the package, it is easy to generate training data for chatbots or question-answering models. The available constructor options provide flexibility for customizing the behavior of the package.

CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2023-05-05

- Initial release

## [0.1.2] - 2023-05-05

-  Proivde code completion in answer values.

## [1.0.0] - 2023-05-07

-  Implement generateFromUrls to allow full scraping through screenshots-pagepixels package of webpage HTML and to generate question:answer pairs and embeddings from the webpage content. Updates constructor to pass an options hash and includes additional refinement options and verbose flag. This version is not backward compatible with previous versions.

## [1.0.1] - 2023-05-07

-  Provides additional clarity and examples to the readme.

## [1.1.0] - 2023-05-07

-  Ensure embedding model configuration is properly set and implement generateEmbeddingForText as a standalone function.

## [2.0.0] - 2023-05-07

-  Make the method names more descriptive and provide additional documentation in readme.  This version is not backward compatible with previous verisons.

## [2.0.1]

-  Fix verbose log.

## [2.0.2]

-  Fix verbose Log

## [2.0.3]

-  Fixed PagePixels dependency version

## [2.0.4]

-  Implement true tokenization for chunks using gpt-3-encoder.

## [2.0.5] 2023-07-14

-  Optimize splitTextIntoChunks method.

embeddings.js
const cheerio = require("cheerio");
const https = require("https");
const GPTEncoder = require('gpt-3-encoder')
const ScreenshotsPagepixels = require("screenshots-pagepixels");

class Embeddings {
  constructor(options) {
    this.apiKey = options.apiKey;
    this.embeddingModel = options.embeddingModel || "text-embedding-ada-002";
    this.completionModel = options.completionModel || "text-davinci-003";
    this.completionModelOptions = options.completionModelOptions || { max_tokens: 2000, n: 1, stop: null, temperature: 0.7 };
    this.screenshotApiKey = options.screenshotApiKey;
    this.screenshotOptions = options.screenshotOptions || {};
    this.chunkMaxTokens = options.chunkMaxTokens || 800;
    this.promptRefinement = options.promptRefinement || "";
    this.verbose = options.verbose || false;
  }

  async openaiCall(prompt, url, model = null) {
    const postData = model === this.embeddingModel
      ? JSON.stringify({
          input: prompt,
          model: this.embeddingModel,
        })
      : JSON.stringify({
          prompt,
          model: this.completionModel,
          ...this.completionModelOptions,
        });
  
    const options = {
      hostname: "api.openai.com",
      path: url,
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(postData),
        Authorization: `Bearer ${this.apiKey}`,
      },
    };
  
    return new Promise((resolve, reject) => {
      const req = https.request(options, (res) => {
        res.setEncoding("utf8");
        let responseBody = "";
        res.on("data", (chunk) => {
          responseBody += chunk;
        });
        res.on("end", () => {
          const response = JSON.parse(responseBody);
  
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve(model ? response.data[0].embedding : response.choices[0].text);
          } else {
            const error = new Error(response.message || "OpenAI API error");
            error.statusCode = res.statusCode;
            error.response = response;
            reject(error);
          }
        });
      });
  
      req.on("error", (error) => {
        console.error("Error calling OpenAI API:", error);
        reject(error);
      });
  
      req.write(postData);
      req.end();
    });
  }  

  extractText(html) {
    if(this.verbose){ console.log("extracting text"); }
    const $ = cheerio.load(html);
    $("script, style, link").remove(); // remove script, style, and link tags
    const text = $('body').text().replace(/\s+/g, ' ').trim(); // extract text only from body element
    if(this.verbose){ console.log(text); }
    return text;
  }

  async generateQaPairsFromText(text, url=undefined) {
    const maxTokens = this.chunkMaxTokens;
    const textChunks = this.splitTextIntoChunks(text, maxTokens);

    let allQaPairs = [];

    for (const chunk of textChunks) {
      if(this.verbose){ console.log("Calling OpenAI to process chunk"); }
      if(this.verbose) console.log("Chunk: ", chunk)
      const prompt = `Process the following text into a list of question-answer pairs associated with the relevant content on the page, like: [{"question": "this is the question", "answer": "this is the answer"}]. ${this.promptRefinement}. Return only valid JSON in your response:\n\n${chunk}\n\nQuestion-answer pairs valid JSON:`;
      const completion = await this.openaiCall(prompt, "/v1/completions", null);
      if(this.verbose){ console.log("Processing Text"); }
      // Escape double quotes only within code examples
      const escapedCompletion = completion.trim().replace(/(```[\s\S]*?```)/g, (match) => {
        return match.replace(/(?<=^```[^`]*)(?<!\\)"/gm, '\\"');
      });
      if(this.verbose){ console.log("Escaped Completion: ", escapedCompletion); }

      // Parse the escaped JSON string
      try{
        let qaPairs = JSON.parse(escapedCompletion);
        
        // Add the URL key to each qaPair
        if(url){
          qaPairs = qaPairs.map(qaPair => ({ ...qaPair, url }));
        }
        if(this.verbose){ console.log(qaPairs); }
        allQaPairs = allQaPairs.concat(qaPairs);
      } catch (error) {
        console.error("Error parsing JSON returned by OpenAI:", error);
      }
    }

    return allQaPairs;
  }
  
  
  splitTextIntoChunks(text, maxTokens) {
    const tokens = GPTEncoder.encode(text)
    const chunks = [];

    for(let token = 0; token < tokens.length; token += maxTokens) {
      chunks.push(
        GPTEncoder.decode(tokens.slice(token, token + maxTokens))
      )
    }
  
    return chunks;
  }

  async generateEmbeddingForText(text) {
    try {
      const embedding = await this.openaiCall(text, "/v1/embeddings", this.embeddingModel);
      return embedding;
    } catch (error) {
      console.error("Error generating embedding:", error);
      return null;
    }
  }

  async generateQaEmbeddingsFromQaPairs(qaPairs) {
    if(this.verbose){ console.log("Generating Embeddings"); }

    // Generate embeddings for all questions
    const embeddings = [];
    for (const qaPair of qaPairs) {
      const question = qaPair.question;
      const embedding = await this.generateEmbeddingForText(question);
      if (embedding) {
        embeddings.push({
          question,
          answer: qaPair.answer,
          url: qaPair.url,
          embedding,
        });
      }
    }

    return embeddings;
  }

  async generateQaEmbeddingsFromText(text) {
    const extractedText = this.extractText(text);
    const qaPairs = await this.generateQaPairsFromText(extractedText);
    const embeddingsResult = await this.generateQaEmbeddingsFromQaPairs(qaPairs);
    return embeddingsResult;
  }

  async generateQaEmbeddingsFromUrls(urls) {
    const pagepixels = new ScreenshotsPagepixels(this.screenshotApiKey);
    const htmlPromises = urls.map(async (url) => {
      if(this.verbose) console.log("processing", url);
      const options = { url, html_only: true, ...this.screenshotOptions };
      const html = await pagepixels.snap(options);
      return JSON.parse(html).html;
    });

    const htmlResults = await Promise.all(htmlPromises);
    const textResults = htmlResults.map((html) => this.extractText(html));
    const qaPairsResults = await Promise.all(textResults.map((text, index) => this.generateQaPairsFromText(text, urls[index])));
    const embeddingsResults = await Promise.all(qaPairsResults.map((qaPairs) => this.generateQaEmbeddingsFromQaPairs(qaPairs)));

    const flatEmbeddingsResults = [].concat(...embeddingsResults);
    return flatEmbeddingsResults;
  }

}

module.exports = Embeddings;
